### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by TesheMaximillan**
## Milestone VIII:
***
> Create integration tests for all of the views used in the project.

### Basic Requirements:
***
* N+1 problem is solved when fetching all posts and their comments for a user

__Integration Tests:__

* User Index page
* User Show page
* User Post Index page
* Post Show page

> N.B: We don't have a login page so making an integration test for it makes the test fail, so, it has to be done after Authentication with devise.

### CODE REVIEW

File: spec/views/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  it 'when I click a user\'s post, I am redirected to the post show page' do
+    click_on 'Post 1'
+    expect(page).to have_current_path user_post_path(@user, @post)
+  end
+
+  it 'When I click to see all posts, I am redirected to the user\'s post index page' do
+    click_on 'Show all user post'
+    expect(page).to have_current_path user_posts_path(@user)
+  end
+end
```
Comment (BenMukebo) [inline spec/views/user_show_spec.rb:43]: - [ ] Great job so far 👏🏻 however you should have implemented the featured test for **Login page**: 👇🏻 

* Tests that returns a detailed error when no `username` and `password` is entered when the submit button is clicked
* Tests that returns a detailed error when an incorrect `username` or `password` is entered when the submit button is clicked
* Tests that redirects to the root page when a correct `username` and `password` is entered.

**See example code below**


```ruby
 describe 'the signin process', type: :feature do
     before :each do
         User.create(email: 'user1@gmail.com', password: 'password', name: 'User1')
     end

     it 'signs @user in' do
         visit '/users/sign_in'
         fill_in 'Email', with: 'user1@gmail.com'
         fill_in 'Password', with: 'password'
         click_button 'Log in'
         expect(current_path).to eq(root_path)
         expect(page).to have_text('Signed in successfully.')
     end

     it 'does not sign in with wrong info' do
         visit '/users/sign_in'
         fill_in 'Email', with: 'user1@gmail.com'
         fill_in 'Password', with: 'mypassword'
         click_button 'Log in'
         expect(page).to have_text('Invalid Email or password.')
     end
 end
```

**Note** Your code does not need to be exact.

**Review by BenMukebo — CHANGES_REQUESTED** [review]
### Hi @TesheMaximillan, @dewslyse,

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

### To highlight 💪🏻

- Gitflow was correctly used.  ✔️   
- Good commit messages ✔️ 
- Your app solves N+1 problem✔️ 
 👍  Good job!


## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_


- [x] This is a group project. Please let's include all the authors in the README file. This makes the README file professional and descriptive. 👨🏻‍🎨 

- [ ] It would be nice to add explicit steps on how to run tests inside your README file. ✍🏻


- The README file It is the first file a person will see when they encounter your project, so it should be fairly brief but detailed. It will make your project stand out from a bunch of others 
## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/user_show_spec.rb (hunk 1 above)
Comment (dewslyse) [reply spec/views/user_show_spec.rb:43]: Thanks @BenMukebo  for the review. Though the requirements request tests for the login page, this has not been a previously requested part of the wireframe build and has yet to be implemented at this stage. Please reference this [**issue**](https://github.com/microverseinc/curriculum-rails/issues/107) that was submitted to correct this requirement for this project

Comment (TesheMaximillan) [reply spec/views/user_show_spec.rb:43]: Hello Reviewer

Also, I did mention this on the pull request.

Comment (TesheMaximillan) [reply spec/views/user_show_spec.rb:43]: Please check through the Readme file, all the authors are mentioned there

Comment (BenMukebo) [reply spec/views/user_show_spec.rb:43]: Hello @TesheMaximillan, @dewslyse
I understand what you said, but as you can notice from the screenshot below 👇🏻 I didn't invent the project requirements, I just followed what is displayed. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

19test

Comment (TesheMaximillan) [reply spec/views/user_show_spec.rb:43]: Hello @BenMukebo;

It's not about whether the code review is fair or not, Microverse is making changes to the curriculum, and we know we have to do the login page, actually, we didn't do the login page and it's not possible to make a test without having the login page.  Thank you for your understanding and please try to see the issue page too. [issue](https://github.com/microverseinc/curriculum-rails/issues/107)

Comment (Whoistolu) [reply spec/views/user_show_spec.rb:43]: Hello @TesheMaximillan , your project is complete.

**Review by Whoistolu — APPROVED** [review]
### Hi @TesheMaximillan 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

# STATUS: APPROVED 🟢
**Congratulations! 🎉** 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @whoistolu in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by RolandM99 — COMMENTED** [review]
Hi @TesheMaximillan and @dewslyse , 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request ❌

You have submitted your PR for an incorrect activity. Instead of submitting for re-review of the milestone about `integration specs and N+1 problem` you submitted your PR for `add devise` activity which you didn't yet implement. Kindly, submit correctly your PR.

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Pull request merged**